warp/Network/Wai/Handler/Warp/WithApplication.hs

```diff
@@ -0,0 +1,94 @@
+
+module Network.Wai.Handler.Warp.WithApplication (
+  withApplication,
+  testWithApplication,
+  openFreePort,
+) where
+
+import           Control.Concurrent
+import           Control.Exception
+import           Network.Socket
+import           Network.Wai
+import           Network.Wai.Handler.Warp.Run
+import           Network.Wai.Handler.Warp.Settings
+import           Network.Wai.Handler.Warp.Types
+
+data App
+  = App {
+    appThread :: ThreadId,
+    appKilled :: Waiter (),
+    appPort :: Int
+  }
+
+-- | Runs the given 'Application' on a free port. Passes the port to the given
+-- operation and executes it, while the 'Application' is running. Shuts down the
+-- server before returning.
+withApplication :: IO Application -> (Port -> IO a) -> IO a
+withApplication mkApp action = do
+  app <- mkApp
+  bracket (acquire app) free (\ runningApp -> action (appPort runningApp))
+  where
+    acquire :: Application -> IO App
+    acquire app = do
+      start <- mkWaiter
+      killed <- mkWaiter
+      thread <- forkIO $ do
+        (port, sock) <- openFreePort
+        let settings =
+              defaultSettings{
+                settingsBeforeMainLoop = notify start port
+              }
+        runSettingsSocket settings sock app
+          `finally` notify killed ()
```

Comment:
And this reminded me of what I'd meant originally about using async for this:

```haskell
app <- mkApp
(port, sock) <- openFreePort
let settings = defaultSettings {...}
res <- race (runSettingsSocket settings sock app) (action port)
case res of
  Left () -> error "Unexpected: runSettingsSocket exited"
  Right x -> return x
```

And after writing _that_ out, I realized: the `openFreePort` call actually needs to be wrapped in a `bracket` to close the Socket afterwards.